accord-core/src/main/java/accord/coordinate/tracking/ReadTracker.java

```diff
@@ -18,232 +18,281 @@
 
 package accord.coordinate.tracking;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.LinkedHashSet;
-import java.util.List;
-import java.util.Set;
-import java.util.function.Function;
-import java.util.function.IntFunction;
+import java.util.*;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
 
+import accord.topology.Shard;
+import accord.topology.ShardSelection;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
 
-import accord.coordinate.tracking.ReadTracker.ReadShardTracker;
 import accord.local.Node.Id;
-import accord.topology.Shard;
 import accord.topology.Topologies;
 
+import static accord.coordinate.tracking.AbstractTracker.ShardOutcomes.*;
 import static com.google.common.collect.Sets.newHashSetWithExpectedSize;
 
-public class ReadTracker<T extends ReadShardTracker> extends AbstractResponseTracker<T>
+public class ReadTracker extends AbstractTracker<ReadTracker.ReadShardTracker, Boolean>
 {
+    private static final ShardOutcome<ReadTracker> DataSuccess = tracker -> {
+        --tracker.waitingOnShards;
+        --tracker.waitingOnData;
+        return ShardOutcomes.Success;
+    };
+
     public static class ReadShardTracker extends ShardTracker
     {
-        private boolean hasData = false;
+        protected boolean hasData = false;
+        protected int acceptQuorum;
         protected int inflight;
-        private int contacted;
-        private int slow;
+        protected int contacted;
+        protected int slow;
 
         public ReadShardTracker(Shard shard)
         {
             super(shard);
         }
 
-        public void recordInflightRead(Id node)
+        public ShardOutcome<? super ReadTracker> recordInFlightRead(boolean ignore)
         {
             ++contacted;
             ++inflight;
+            return NoChange;
         }
 
-        // TODO: this is clunky, restructure the tracker to handle this more cleanly
-        // record a node as contacted, even though it isn't
-        public void recordContacted(Id node)
+        public ShardOutcome<? super ReadTracker> recordSlowResponse(boolean ignore)
         {
-            ++contacted;
+            Preconditions.checkState(!hasFailed());
+            ++slow;
+
+            if (shouldRead() && canRead())
+                return SendMore;
+
+            return NoChange;
```

Comment:
Slow means we haven't received a prompt response? It means we still expect another response (either timeout or success or whatever).

Since this is not a terminal response, being unable to read isn't a failure, no.